File: Taller4/Taller4.py
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from scipy import signal

# 1. Definición de Parámetros del Circuito LTI
R = 100.0
L = 0.1
Cap = 1e-6

A = [[0, 1/Cap], 
     [-1/L, -R/L]]
B = [[0], 
     [1/L]]
C = [[1, 0]]
D = [[0]]

# Creación del objeto de estado con SciPy
sys_rlc = signal.StateSpace(A, B, C, D)

# 2. Generación del Archivo CSV (Señal Arbitraria Ficticia)
t_generado = np.linspace(0, 0.06, 600)
# Vamos a hacer una señal que empiece en 0V, salte a 8V y luego baje a 3V
u_generada = np.piecewise(t_generado, 
                          [t_generado < 0.015, (t_generado >= 0.015) & (t_generado < 0.04), t_generado >= 0.04], 
                          [0, 8, 3])

df = pd.DataFrame({'tiempo': t_generado, 'voltaje': u_generada})
df.to_csv('senal_arbitraria.csv', index=False)
print("Archivo 'senal_arbitraria.csv' creado exitosamente.")

# 3. Lectura de CSV y Simulación LTI (El reto en sí)
datos_leidos = pd.read_csv('senal_arbitraria.csv')
t_in = datos_leidos['tiempo'].values
u_in = datos_leidos['voltaje'].values

# Función lsim de scipy.signal simula la respuesta de sistemas de tiempo continuo
tout, yout, xout = signal.lsim(sys_rlc, U=u_in, T=t_in)

# 4. Graficar los resultados
plt.figure(figsize=(8, 5))
plt.plot(t_in, u_in, 'r--', label='Voltaje de Entrada (CSV)')
plt.plot(tout, yout, 'b-', linewidth=2, label='Voltaje del Capacitor (Simulado)')
plt.title('Simulación en Python usando scipy.signal y archivo CSV')
plt.xlabel('Tiempo [s]')
plt.ylabel('Voltaje [V]')
plt.legend()
plt.grid(True)
plt.show()

File: Taller4/taller4.m

% ==========================================
% PRÁCTICA 4 - CONTROL MODERNO
% ==========================================
pkg load control
pkg load signal
clear all; clc; close all;

% A partir de aquí va el resto de tu código...

R=100;
L=0.1;
Cap =1e-6;
A= [0 1/Cap; -1/L -R/L];
B= [0;1/L];
C=[1 0];
D=0;
D3=[0; 0];
C2=[0 1];
C3=[1 0;0 1];

[num,den] = ss2tf(A,B,C,D);
tf1=tf(num,den);



[num2,den2] = ss2tf(A,B,C2,D);
tf2=tf(num2,den2);
subplot(2,1,1);


step(tf1);
ylabel("Vc");
subplot(2,1,2);
step(tf2);
ylabel("I(t)");

%%
figure
sys = ss(A,B,C,D);
subplot(2,1,1);
step(sys);
ylabel("Vc");
subplot(2,1,2);
impulse(sys);
ylabel("I(t)");


%%
[num, den]= ss2tf(A,B,C,D);
tf1= tf(num, den);
[num2, den2]= ss2tf(A,B,C2,D);
tf2= tf(num2, den2);

figure
subplot(2,1,1);
step(tf1);
ylabel("Vc");
subplot(2,1,2);
step(tf2);
ylabel("I(t)");


%%
[num, den]= ss2tf(A,B,C,D);
tf1= tf(num, den);
[num2, den2]= ss2tf(A,B,C2,D);
tf2= tf(num2, den2);

figure
subplot(2,1,1);
step(tf1);
ylabel("Vc");
subplot(2,1,2);
step(tf2);
ylabel("I(t)");

%%
ts = 0.015;
u = 1;
x0 = [0; 0];
[t, X] = ode45(@(t, x) modelRLC(t, x, A, B, u), [0 ts], x0);

Vc = X(:, 1);
I = X(:, 2);

figure;

C = [1 0;0 1];
D = 0;
Y = (C * X')';


subplot(2,1,1);
plot(t, Y(:,1), 'LineWidth', 2);
grid on;
title('Respuesta del sistema usando ode45');
ylabel('Vc [V]');
xlim([0 0.015]);



subplot(2,1,2);
plot(t, Y(:,2), 'LineWidth', 2);
grid on;
xlabel('Time [s]');
ylabel('I(t) [A]');
xlim([0 0.015]);



function dx = modelRLC(t, x, A, B, u)
    dx = A * x + B * u;             % Ecuación de Estado
end
